Docs: Plugin E2E: Getting started docs (#701)
grafana/plugin-tools: docusaurus/docs/e2e-test-a-plugin/getting-started.md

---
id: getting-started

> title: Getting started

Reviewer: ```suggestion
title: Get started
```
nit: consistency

> description: Getting started

Reviewer: ```suggestion
description: Get started
```
nit: consistency

keywords:
  - grafana
  - plugins
  - plugin
  - testing
  - e2e

>   - getting-started

Reviewer: ```suggestion
  - getting-started
  - get-started
```

import ScaffoldPluginE2EStartGrafanaYarn from '@snippets/plugin-e2e-start-grafana.yarn.md';
import ScaffoldPluginE2ERunTestsNPM from '@snippets/plugin-e2e-run-tests.npm.md';
import ScaffoldPluginE2ERunTestsPNPM from '@snippets/plugin-e2e-run-tests.pnpm.md';
import ScaffoldPluginE2ERunTestsYarn from '@snippets/plugin-e2e-run-tests.yarn.md';
import ScaffoldPluginE2EDSWorkflowNPM from '@snippets/plugin-e2e-ds-workflow.npm.md';
import ScaffoldPluginE2EDSWorkflowYarn from '@snippets/plugin-e2e-ds-workflow.yarn.md';
import ScaffoldPluginE2EDSWorkflowPNPM from '@snippets/plugin-e2e-ds-workflow.pnpm.md';
import ScaffoldPluginE2EProjectsConfigNPM from '@snippets/plugin-e2e-auth-project.npm.md';
import ScaffoldPluginE2EProjectsConfigYarn from '@snippets/plugin-e2e-auth-project.yarn.md';
import ScaffoldPluginE2EProjectsConfigPNPM from '@snippets/plugin-e2e-auth-project.pnpm.md';

# Getting started

This article will guide through how to install and configure `@grafana/plugin-e2e`, write tests and setup a basic Github workflow that can run your e2e tests targeting multiple versions of Grafana.

## Prerequisites

> - You need to have a Grafana plugin [development environment](https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment)

Reviewer: ```suggestion
- You need to have a suitable Grafana development environment where you can run your plugin, we recommend using the [docker development environment](https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment) included with `create-plugin`
```

Author: This is not entirely true - a Grafana development environment is not required. 

What about something like this: 

- You need to have a Grafana plugin [development environment](https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment)
- Grafana running on your local machine. We recommend using docker development environment included in plugin development environment, but you may also build Grafana from [source](https://github.com/grafana/grafana/blob/main/contribute/developer-guide.md).

- Node.js 18+
- Docker
- Basic Knowledge of Playwright. If you have not worked with Playwright before, we recommend following the [Getting started](https://playwright.dev/docs/intro) section in their documentation.

### Installing Playwright

> Please refer to the [Playwright documentation](https://playwright.dev/docs/intro#installing-playwright) for instruction on how to install. `@grafana/plugin-e2e` extends Playwright APIs, so you need to have `Playwright/test` with a minimum version of 0.40.0 installed as a dev dependency in the package.json file of your plugin.

Reviewer: ```suggestion
Please refer to the [Playwright documentation](https://playwright.dev/docs/intro#installing-playwright) for instruction on how to install. `@grafana/plugin-e2e` extends Playwright APIs, so you need to have `Playwright/test` with a minimum version of 0.40.0 installed as a dev dependency in the package.json file of your plugin.

After installing, you should have (at least) a `playwright.config.ts` file in the project root folder for your plugin.
```

## Set up `plugin-e2e`

### Step 1: Installing @grafana/plugin-e2e

Now open the terminal and run the following command in your plugin's project directory:

<CodeSnippets
snippets={[
{ component: ScaffoldPluginE2InstallNPM, label: 'npm' },
{ component: ScaffoldPluginE2EInstallYarn, label: 'yarn' },
{ component: ScaffoldPluginE2EInstallPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>

### Step 2: Configure Playwright

Open the Playwright config file that was generated when Playwright was installed.

1. Uncomment `baseUrl` and change it to `'http://localhost:3000'`.

```ts title="playwright.config.ts"
  baseURL: 'http://localhost:3000',
```

> 2. Playwright uses [projects](https://playwright.dev/docs/test-projects) to logically group tests that have the same configuration. We're going to add two projects. The first project called `auth` is a setup project will login to Grafana and store the authenticated state on disk. The second project `run-tests` runs all the tests in a browser of choice. By adding a dependency to the `auth` project we ensure login only happens once, and all tests in the `run-tests` project will start already authenticated.

Reviewer 2: Testing this out locally and I see playwright adds a /playwright/.auth/ directory to my plugin project. Should we .gitignore that directory?

Author: I believe it's added to the .gitignore file when you run `npm init playwright@latest`

Reviewer 2: Running `npm init playwright@latest` gives me the following.

![image](https://github.com/grafana/plugin-tools/assets/73201/e81f8f17-a19c-4c98-ada8-680d1b3fff0a)

I'm under the impression `.auth` directory is created by the plugin-e2e auth project config:

```
use: { ...devices['Desktop Chrome'], storageState: 'playwright/.auth/user.json' },
```

I'm not very familar with playwright though. Maybe it should live under `.cache`?

Author: Ah I see - no it's meant to live under `playwright/.auth/`. I wonder why it's not added when installing playwright. 

We could add it to this getting started guide, but would prefer to keep it as short and simple as possible. Do you think adding `playwright/.auth/` to the .gitgnore file as part of the create-plugin scaffolding would be sufficient?

> 2. Playwright uses [projects](https://playwright.dev/docs/test-projects) to logically group tests that have the same configuration. We're going to add two projects. The first project called `auth` is a setup project will login to Grafana and store the authenticated state on disk. The second project `run-tests` runs all the tests in a browser of choice. By adding a dependency to the `auth` project we ensure login only happens once, and all tests in the `run-tests` project will start already authenticated.

Reviewer: ```suggestion
2. Playwright uses [projects](https://playwright.dev/docs/test-projects) to logically group tests that have the same configuration. We're going to add two projects:
   1. `auth` is a setup project will login to Grafana and store the authenticated state on disk.
   1. `run-tests` runs all the tests in a browser of choice. By adding a dependency to the `auth` project we ensure login only happens once, and all tests in the `run-tests` project will start already authenticated.
```

optional - think it improves readability

Author: Yep I like it!

   Your Playwright config should have the following project configuration:

<CodeSnippets
snippets={[
{ component: ScaffoldPluginE2EProjectsConfigNPM, label: 'npm' },
{ component: ScaffoldPluginE2EProjectsConfigYarn, label: 'yarn' },
{ component: ScaffoldPluginE2EProjectsConfigPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>

### Step 3: Start Grafana

Next, start up the latest version of Grafana on your local machine.

<CodeSnippets
snippets={[
{ component: ScaffoldPluginE2EStartGrafanaNPM, label: 'npm' },
{ component: ScaffoldPluginE2EStartGrafanaYarn, label: 'yarn' },
{ component: ScaffoldPluginE2EStartGrafanaPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>

If you want to start a specific version of Grafana, you can do that by specifying the `GRAFANA_VERSION` environment variable.

```bash
GRAFANA_VERSION=10.1.6 npm run server
```

### Step 4: Write tests

> In this example, we're using the panel edit page to test the query editor for a backend data source plugin. The plugin was scaffolded with the [create-plugin](https://github.com/grafana/plugin-tools/tree/main/packages/create-plugin) tool, and for this data source the query endpoint returns hard coded data points. This test asserts that the values `1` and `3` are being displayed in the `Table` panel.

Reviewer: When following a getting started guide, I like if I can see everything run through successfully before adding to my own project. If this example is just based on the default backend DS template, wdyt about adding an explanation near the beginning where we advise, if you would like to follow along with our example tests, use create-plugin to generate a backend datasource?

Author: Yeah I like that! See [this](https://github.com/grafana/plugin-tools/pull/701/commits/f1e678d7b440f662d969eeb398bd828ae2a4d0f2) commit

Reviewer: Added a comment

> In this example, we're using the panel edit page to test the query editor for a backend data source plugin. The plugin was scaffolded with the [create-plugin](https://github.com/grafana/plugin-tools/tree/main/packages/create-plugin) tool, and for this data source the query endpoint returns hard coded data points. This test asserts that the values `1` and `3` are being displayed in the `Table` panel.

Reviewer: ```suggestion
You are now ready to write your tests. In this example, we're using the panel edit page to test the query editor for a backend data source plugin. The plugin was scaffolded with the [create-plugin](https://github.com/grafana/plugin-tools/tree/main/packages/create-plugin) tool, and for this data source the query endpoint returns hard coded data points. This test asserts that the values `1` and `3` are being displayed in the `Table` panel.
```

Reviewer: we could also optionally run the login/auth test prior to this to verify everything is working as expected with the setup?

```ts title="queryEditor.spec.ts"
import { test, expect } from '@grafana/plugin-e2e';

test('data query should return values 1 and 3', async ({ panelEditPage, readProvisionedDataSource }) => {
  const ds = await readProvisionedDataSource({ fileName: 'datasources.yaml' });
  await panelEditPage.datasource.set(ds.name);
  await panelEditPage.setVisualization('Table');
  await expect(panelEditPage.refreshPanel()).toBeOK();
  await expect(panelEditPage.panel.data).toContainText(['1', '3']);
});
```

### Step 5: Run tests

> Now you open a new terminal and run the test script from within your local plugin development directory.

Reviewer: ```suggestion
Open a new terminal and run the test script from within your local plugin development directory.
```

<CodeSnippets
snippets={[
{ component: ScaffoldPluginE2ERunTestsNPM, label: 'npm' },
{ component: ScaffoldPluginE2ERunTestsYarn, label: 'yarn' },
{ component: ScaffoldPluginE2ERunTestsPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>

### Step 6: Run tests in CI

> The following workflow can be used to run e2e tests against a matrix of Grafana versions for every PR in your Github repository. Note that this is a generic example based on a backend plugin. You may want to alter or remove a few of the steps in the `playwright-tests` job before using it in your plugin.

Reviewer: ```suggestion
We recommend updating your CI workflow to include the running of e2e tests to continuously check for breakages. The following workflow can be used to run e2e tests against a matrix of Grafana versions for every PR in your Github repository. Note that this is a generic example based on a backend plugin. You may want to alter or remove a few of the steps in the `playwright-tests` job before using it in your plugin.
```

Reviewer: We need to explain the "matrix of grafana versions" more.

Author: Yeah let's just say `a range of Grafana versions`, and then we can elaborate on this in the full `CI` article

Reviewer: sounds good

]}
groupId="package-manager"
queryString="current-package-manager"
/>

```ts title="playwright.config.ts"
projects: [
  {
    name: 'auth',
    testDir: 'node_modules/@grafana/plugin-e2e/dist/auth',
    testMatch: [/.*\.js/],
  }
```

### Step 3: Start Grafana

Next, start up the latest version of Grafana on your local machine.

<CodeSnippets
snippets={[
{ component: ScaffoldPluginE2EStartGrafanaNPM, label: 'npm' },
{ component: ScaffoldPluginE2EStartGrafanaYarn, label: 'yarn' },
{ component: ScaffoldPluginE2EStartGrafanaPNPM, label: 'pnpm' }
]}
groupId="package-manager"
queryString="current-package-manager"
/>

If you want to start a specific version of Grafana, you can do that by specifying the `GRAFANA_VERSION` environment variable.

```bash
GRAFANA_VERSION=10.1.6 npm run server
```

### Step 4: Write tests

In this example, we're using the panel edit page to test a data source plugin. When the provided query is valid, the response status code is expected to be in the range 200-299.

```ts title="queryEditor.spec.ts"
import { test, expect } from '@grafana/plugin-e2e';

test('data query should return values 1 and 3', async ({ panelEditPage, readProvisionedDataSource }) => {

>   const ds = await readProvisionedDataSource({ fileName: 'datasources.yaml' });

Author: The [pr](https://github.com/grafana/plugin-tools/pull/740) that adds this fixture is not merged yet

---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Getting started

This article will guide through how to install and configure `@grafana/plugin-e2e`, write tests and setup a basic Github workflow that can run your e2e tests targeting multiple versions of Grafana.

## Prerequisites

- You need to have a Grafana plugin [development environment](https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment)
- Node.js 18+
- Docker
- Basic Knowledge of Playwright. If you have not worked with Playwright before, we recommend following the [Getting started](https://playwright.dev/docs/intro) section in their documentation.

### Installing Playwright

Please refer to the [Playwright documentation](https://playwright.dev/docs/intro#installing-playwright) for instruction on how to install. `@grafana/plugin-e2e` extends Playwright APIs, so you need to have `Playwright/test` with a minimum version of 0.40.0 installed as a dev dependency in the package.json file of your plugin.

## Set up `plugin-e2e`

### Step 1: Installing @grafana/plugin-e2e

Now open the terminal, cd into your plugin root folder and install `@grafana/plugin-e2e`.

> <Tabs

Author: Ah I was not aware of this - this is great! Will fix.

defaultValue="npm">
<TabItem value="npm">

```bash
npm install @grafana/plugin-e2e@latest --save-dev
```

</TabItem>

<TabItem value="yarn">

```bash
yarn add @grafana/plugin-e2e@latest --dev
```

</TabItem>

<TabItem value="pnpm">

```bash
pnpm add @grafana/plugin-e2e@latest -D
```

</TabItem>
</Tabs>

### Step 2: Configure Playwright

Open the `playwright.config.[js|ts]` file that was generated when Playwright was installed.

1. Change the `baseUrl` to `'http://localhost:3000'`.

```ts title="playwright.config.ts"
use: {
  baseURL: 'http://localhost:3000',
  ...
}
```

2. Add a new `auth` setup project. This will login to Grafana and store the authenticated state on disk.

```ts title="playwright.config.ts"
projects: [
  {
    name: 'auth',
    testDir: 'node_modules/@grafana/plugin-e2e/dist/auth',
    testMatch: [/.*\.js/],
  }
```

3. Then add another project that runs all the tests in a browser of choice. This project needs to depend on the `auth` project we added in the previous step. That will ensure login only happens once, and all tests in the `run-tests` project will start already authenticated.

```ts title="playwright.config.ts"
projects: [
  {
    name: 'auth',

>     testDir: 'node_modules/@grafana/plugin-e2e/dist/auth',

Reviewer 2: This path might not be correct for PNPM due to its symlink strategy. It's a shame testDir doesn't resolve npm packages automagically. One solution here would be to use `require.resolve` to get the path...

```ts
const pluginE2eAuth = `${dirname(require.resolve('@grafana/plugin-e2e'))}/auth`;

// defineConfig bits...
 projects: [
    // 1. Login to Grafana and store the cookie on disk for use in other tests.
    {
      name: 'auth',
      testDir: pluginE2eAuth,
      testMatch: [/.*\.js/],
    },
```

Author: Works like a charm, thank you!

## Set up `plugin-e2e`

> > If you would like to follow along with our example tests, use the [create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool to generate a backend Data Source plugin.

Reviewer: We can pass args now so could probably give this as a single command. 

i.e. `npx @grafana/create-plugin@latest  --pluginName='Sample datasource' --orgName='sample-org' --pluginDescription='This is a sample datasource.' --pluginType='datasource' --hasBackend`
